mempool/clist_mempool.go

```diff
@@ -780,3 +784,42 @@ func (rc *recheck) setRecheckFull() bool {
 func (rc *recheck) consideredFull() bool {
 	return rc.recheckFull.Load()
 }
+
+// CListIterator implements an Iterator that traverses the CList sequentially. When the current
+// entry is removed from the mempool, the iterator starts from the beginning of the CList. When it
+// reaches the end, it waits until a new entry is appended.
+type CListIterator struct {
+	txs    *clist.CList    // to wait on and retrieve the first entry
+	cursor *clist.CElement // pointer to the current entry in the list
+}
+
+func (mem *CListMempool) NewIterator() Iterator {
+	return &CListIterator{
+		txs: mem.txs,
+	}
+}
+
+// WaitNextCh returns a channel to wait for the next available entry. The channel will be closed
+// once the entry is added to the channel or when reaching the end of the list.
+func (iter *CListIterator) WaitNextCh() <-chan Entry {
+	ch := make(chan Entry)
+	// Spawn goroutine that waits for the next entry, saves it locally, and puts it in the channel.
+	go func() {
+		if iter.cursor == nil {
+			// We are at the beginning of the iteration or the saved entry got removed: wait until
+			// the list becomes not empty and select the first entry.
+			<-iter.txs.WaitChan()
+			iter.cursor = iter.txs.Front()
+		} else {
+			// Wait for the next entry after the current one.
+			<-iter.cursor.NextWaitChan()
+			iter.cursor = iter.cursor.Next()
```

Comment:
What if the TX that `iter.cursor` points to got removed while we were blocked in line 815?

Reply:
Hmm, I think the answer to this is [here](https://github.com/cometbft/cometbft/pull/3459/files#diff-3e6afafe7994a696d508a5a5b093343bdb5d1ff2ff00bab0318d2c87547e74e5R228)?

Reply:
Exactly. It will set cursor to nil and start from the beginning on the next iteration.